src/string.rs

```diff
@@ -0,0 +1,194 @@
+    }
+
+    fn det_matcher(patterns: Vec<StringPattern>) -> StringManyMatcher {
+        ManyMatcher::from_patterns_with_det_heuristic(patterns, |_| true)
+    }
+
+    pub(super) const MATCHER_FACTORIES: &[fn(Vec<StringPattern>) -> StringManyMatcher] =
+        &[non_det_matcher, default_matcher, det_matcher];
+
+    /// Comparing two lists of pattern matches.
+    ///
+    /// We cannot use equality for two reasons:
+    ///  - we ignore the ordering of the matches
+    ///  - not all keys may be bound depending on the order things were matched
+    ///
+    /// The simplest way to identify a match is with the pair (PatternID, start_pos).
+    pub(super) fn pattern_match_eq(
+        expected: &[PatternMatch<HashMap<StringIndexKey, StringPosition>>],
+        actual: &[PatternMatch<HashMap<StringIndexKey, StringPosition>>],
+    ) -> bool {
+        let get_match_key =
+            |PatternMatch {
+                 pattern,
+                 match_data,
+             }: &PatternMatch<HashMap<StringIndexKey, StringPosition>>| {
+                let Some((StringIndexKey(offset), StringPosition(pos))) = match_data.iter().next()
+                else {
+                    panic!("Empty match data");
+                };
+                (*pattern, pos - offset)
+            };
+        // Build a map from pattern ID to all positions where it matches
+        let get_slice_key = |matches: &[PatternMatch<_>]| {
+            let mut keys = matches.iter().map(get_match_key).fold(
+                HashMap::<_, Vec<_>>::default(),
+                |mut acc, (pattern, pos)| {
+                    acc.entry(pattern).or_default().push(pos);
+                    acc
+                },
+            );
+            keys.values_mut().for_each(|vec| vec.sort());
+            keys
+        };
+        get_slice_key(expected) == get_slice_key(actual)
+    }
+
+    #[test]
+    fn test_string_matching() {
+        let p1 = StringPattern::parse_str("ab$xcd$x");
+        let p2 = StringPattern::parse_str("abcc");
+
+        let matcher = StringManyMatcher::from_patterns(vec![p1, p2]);
+
+        let result = matcher.find_matches(&"abccdc".to_string()).collect_vec();
+
+        assert_eq!(result.len(), 2);
+    }
+
+    #[rstest]
+    #[case("h", vec!["h$aa", "$a$b"])]
+    #[case("aa", vec!["a", "b"])]
+    #[case("aa", vec!["a", "$c$c"])]
+    #[case("haa", vec!["$aaa", "$a$b$c"])]
+    #[case("xag", vec!["x$ag", "$aa"])]
+    #[case("aaaayaazz", vec!["$a$baya", "$a$b$c$d$ez"])]
+    #[case("aea", vec!["$d$aa", "$aea"])]
+    #[case("eaaa", vec!["a", "$aa", "$b$a$ca", "$b$c$e$e"])]
+    #[case("eaa", vec!["a$b$a", "$a$e$e", "b$c$a", "$c$aa"])]
+    fn proptest_fail_cases(#[case] subject: &str, #[case] patterns: Vec<&str>) {
+        use itertools::Itertools;
+
+        let patterns = patterns
+            .into_iter()
+            .map(StringPattern::parse_str)
+            .collect_vec();
+        let all_matches = MATCHER_FACTORIES
```

Comment:
Make a static/helper method for this (passing in `patterns` and `subject`), you use it in proptest too

Reply:
done!